src/datumaro/components/prune.py

```diff
@@ -0,0 +1,323 @@
+# Copyright (C) 2023 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+import logging as log
+import math
+import random
+from abc import ABC, abstractmethod
+from typing import List
+
+import numpy as np
+from sklearn.cluster import KMeans
+
+import datumaro.plugins.ndr as ndr
+from datumaro.components.annotation import HashKey, Label, LabelCategories
+from datumaro.components.dataset import Dataset
+from datumaro.components.explorer import HashInference
+from datumaro.util.hashkey_util import (
+    calculate_hamming,
+    format_templates,
+    select_uninferenced_dataset,
+    templates,
+)
+
+
+def match_num_item_for_cluster(ratio, dataset_len, cluster_num_item_list):
+    total_num_selected_item = math.ceil(dataset_len * ratio)
+
+    cluster_weights = np.array(cluster_num_item_list) / sum(cluster_num_item_list)
+    norm_cluster_num_item_list = (cluster_weights * total_num_selected_item).astype(int)
+    remaining_items = total_num_selected_item - sum(norm_cluster_num_item_list)
+
+    if remaining_items > 0:
+        zero_cluster_indexes = np.where(norm_cluster_num_item_list == 0)[0]
+        add_clust_dist = np.sort(cluster_weights[zero_cluster_indexes])[::-1][:remaining_items]
+
+        for dist in set(add_clust_dist):
+            indices = np.where(cluster_weights == dist)[0]
+            for index in indices:
+                norm_cluster_num_item_list[index] += 1
+
+    elif remaining_items < 0:
+        diff_num_item_list = np.argsort(cluster_weights - norm_cluster_num_item_list)
+        for diff_idx in diff_num_item_list[: abs(remaining_items)]:
+            norm_cluster_num_item_list[diff_idx] -= 1
+
+    return norm_cluster_num_item_list.tolist()
+
+
+class PruneBase(ABC):
+    @abstractmethod
+    def base(self):
```

Comment:
```suggestion
    def base(self, ratio, num_centers, labels, database_keys, item_list, source) -> Tuple[List[DatasetItem], Optional[???]]:
```
Also add typing for the function input arguments